cpp/ql/src/Security/CWE/CWE-570/IncorrectAllocationErrorHandling.ql

```diff
@@ -153,12 +153,39 @@ predicate exprMayThrow(Expr e) {
   )
 }
 
+/** The `std::nothrow_t` class and its `bsl` variant. */
+class NoThrowType extends Struct {
+  NoThrowType() { this.hasGlobalOrStdOrBslName("nothrow_t") }
+}
+
 /** An allocator that might throw an exception. */
 class ThrowingAllocator extends Function {
   ThrowingAllocator() {
     exists(NewOrNewArrayExpr newExpr |
       newExpr.getAllocator() = this and
-      functionMayThrow(this)
+      // Exclude custom overloads of `operator new`.
+      // What we really want here is to only include the functions that satisfy `functionMayThrow`, but
+      // there seems to be examples where `throw()` isn't extracted (which causes false positives).
+      //
+      // As noted in the QLDoc for `Function.getAllocatorCall`:
+      //
+      // "As a rule of thumb, there will be an allocator call precisely when the type
+      // being allocated has a custom `operator new`, or when an argument list appears
+      // after the `new` keyword and before the name of the type being allocated.
+      //
+      // In particular note that uses of placement-new and nothrow-new will have an
+      // allocator call."
+      //
+      // So we say an allocator might throw if:
+      // 1. It doesn't have a body
+      // 2. there is a parameter that is not `nothrow`
+      // 3. the allocator isn't marked with `throw()` or `noexcept`.
+      not exists(this.getBlock()) and
+      exists(Parameter p | p = this.getAParameter() |
+        not p.getUnspecifiedType() instanceof NoThrowType
+      ) and
```

Comment:
Shouldn't 2 be the other way around: require that there _isn't_ a nothrow parameter? Otherwise we'll classify the nothrow-and-aligned allocators as potentially throwing. These are (7) and (8) on https://en.cppreference.com/w/cpp/memory/new/operator_new.

Reply:
Good point. Fixed in 175fdbb1051759e1336971f8aadbc2d6dc51841b. The changes to LGTM results are the same: https://lgtm.com/query/6831593082941660772/